client/components/tinymce/plugins/media/drop-zone.jsx

```diff
@@ -107,18 +144,67 @@ export default React.createClass( {
 		if ( isSingleImage && ! MediaValidationStore.hasErrors( site.ID ) ) {
 			// For single image upload, insert into post content, blocking save
 			// until the image has finished upload
-			if ( selectedItems[ 0 ].transient ) {
+			let selectedItem = selectedItems[ 0 ]
+			if ( selectedItem.transient ) {
 				PostActions.blockSave( 'MEDIA_MODAL_TRANSIENT_INSERT' );
 			}
 
-			onInsertMedia( markup.get( selectedItems[ 0 ] ) );
-			MediaActions.setLibrarySelectedItems( site.ID, [] );
-		} else {
-			// In all other cases, show the media modal list view
-			onRenderModal( { visible: true } );
+			analytics.mc.bumpStat( 'editor_upload_via', 'editor_drop' );
+
+			if ( config.isEnabled( 'post-editor/image-reorientation' ) ) {
+				let handleImageOrientation = ( orientation ) => {
+					// If the orientation is `1` (i.e., "top"), then no further orientation
+					// changes are necessary.
+					if ( orientation > 1 ) {
+						debug( 'Image orientation: %d', orientation );
+
+						return changeImageOrientation( selectedItem.URL, orientation, {
+								maxWidth: editor.getBody().clientWidth
+							} )
+							.then( url => {
+								selectedItem.URL = url;
```

Comment:
Still working my way through the logic here, but at a glance, this mutation of the media state would seem to violate the unidirectional data flow of Flux: https://facebook.github.io/flux/docs/overview.html#structure-and-data-flow

Could we return a new object without mutating the original?

```js
return Object.assign( {}, selectedItem, { URL: url } );
```